build(docs-infra): align code with new CLI v12 apps

gkalpak opened the pull request:
This PR aligns the angular.io config files more closely to how a newly generated CLI v12 app would look like. This helps validate the setup and makes it easier to apply new chages in the future (by preventing the angular.io layout from deviating too much from the default new app layout).

mary-poppins commented:
You can preview aa849d7 at https://pr42358-aa849d7.ngbuilds.io/.

mary-poppins commented:
You can preview 0770f80 at https://pr42358-0770f80.ngbuilds.io/.

gkalpak commented on goldens/size-tracking/aio-payloads.json:
```diff
@@ -2,18 +2,18 @@
-        "main-es2015": 451314,
-        "polyfills-es2015": 55210
+        "runtime-es2017": 4528,
+        "main-es2017": 451652,
```
From 451415 B [on master][1].

[1]: https://circleci.com/gh/angular/angular/997690

gkalpak commented on goldens/size-tracking/aio-payloads.json:
```diff
@@ -2,18 +2,18 @@
-        "main-es2015": 451493,
-        "polyfills-es2015": 55291
+        "runtime-es2017": 4528,
+        "main-es2017": 451801,
```
From 451564 B [on master][1].

[1]: https://circleci.com/gh/angular/angular/997705

mary-poppins commented:
You can preview 0b791f4 at https://pr42358-0b791f4.ngbuilds.io/.

gkalpak commented on aio/tsconfig.json:
```diff
@@ -16,7 +16,7 @@
     "moduleResolution": "node",
     "importHelpers": true,
-    "target": "es2015",
+    "target": "es2017",
```
**Question:**

Do we want to switch to ES2017, which is what CLI v12 apps use by default?
(FYI, it increases the `main` bundle size by 237 B (0.05%), which is a little unexpected, but it could theoretically improve performance.)

Are there any browsers that would support ES2015 but not ES2017?

petebacondarwin replied:
Here appears to be the [support matrix for ES2017 features](https://caniuse.com/async-functions,object-values,object-entries,mdn-javascript_builtins_object_getownpropertydescriptors,pad-start-end,mdn-javascript_grammar_trailing_commas_trailing_commas_in_functions)

It looks like IE is obviously a no-go but then we have downlevelling for that to ES5 already.
Other than that it just looks like Opera mini and Baidu are the only problematic ones, which I think we can ignore.

petebacondarwin replied:
What happens about `async-await` (and the lack of zone.js support) in this case?
Do we just need to ensure that our application code does not use this syntax?

petebacondarwin replied:
It looks like we only use `async-await` in the `DocViewerComponent`, and only for debug logging, which should not require zone.js integration. See `aio/src/app/layout/doc-viewer/doc-viewer.component.ts`.

So I think we are fine either way here.

alfaproject replied:
There is a babel transformer just for async/await in Angular CLI

petebacondarwin replied:
Indeed you are correct! For example:

```ts
        function findCachesAndPrintEntries(swCacheNames, nameSuffix, includeValues, ignoredKeys = []) {
            return (0,tslib__WEBPACK_IMPORTED_MODULE_15__.__awaiter)(this, void 0, void 0, function* () {
                const cacheNames = swCacheNames.filter(name => name.endsWith(nameSuffix));
                for (const cacheName of cacheNames) {
                    const cacheEntries = yield getCacheEntries(cacheName, includeValues, ignoredKeys);
                    yield printCacheEntries(cacheName, cacheEntries);
                }
            });
        }
```

gkalpak replied:
> Other than that it just looks like Opera mini and Baidu are the only problematic ones, which I think we can ignore.

Opera mini and Baidu also [do not support ESM][1], so they should fall back to ES5 anyway.

[1]: https://caniuse.com/es6-module

IgorMinar commented:
@petebacondarwin CLI will downlevel async/await to promises automatically if the target is set to ES2017.

IgorMinar reviewed (approved):
LGTM! thank you @gkalpak for aligning these configs

IgorMinar reviewed (approved):
LGTM 

Reviewed-for: size-tracking

petebacondarwin reviewed (approved):
Needs a rebase by I am happy to say it LGTM:

Reviewed-for: size-tracking

gkalpak committed:
build(docs-infra): align code with new CLI v12 apps

This commit aligns the angular.io config files more closely to how a
newly generated CLI v12 app would look like. This helps validate the
setup and makes it easier to apply new chages in the future (by
preventing the angular.io layout from deviating too much from the
default new app layout).

mary-poppins commented:
You can preview 173e01a at https://pr42358-173e01a.ngbuilds.io/.

mary-poppins commented:
You can preview d55559b at https://pr42358-d55559b.ngbuilds.io/.

gkalpak commented:
Merged as part of #42259.
